### PR TITLE
Test: GET program api endpoint

### LINKED ISSUE #126: Test: GET /organizations/{organization_id}/programs/{program_id} endpoint
## Description
As a developer,
I need write test cases for GET /organizations/{organization_id}/programs/{program_id} endpoints,
so that I can make sure that the endpoint works as it should be.

## Mocks
NA

### Update [Required]
- [ ] test get program with successful outcome (response 200)
- [ ] test get program with at least 2 error responses from 4XX

### Enhancement to Update [Optional]
NA

## Definition of Done
- [ ] All of the required items are completed.
- [ ] Approval by 1 mentor.

## Estimation
1 day

### PR DESCRIPTION
**Pull request by mtreacy002**
### Description
Added test cases for GET /organizations/{organization_id}/programs/{program_id}

Fixes #126 

### Type of Change:
- Code
- Quality Assurance

**Code/Quality Assurance Only**
- New feature (non-breaking change which adds functionality pre-approved by mentors)

### How Has This Been Tested?
- run `python -m unittest discover tests` and all test passed

### Checklist:
- [ ] My PR follows the style guidelines of this project
- [ ] I have performed a self-review of my own code or materials
- [ ] I have commented my code or provided relevant documentation, particularly in hard-to-understand areas

**Code/Quality Assurance Only**
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by mtreacy002** [comment]
Update @anitab-org/bridgeintech-maintainers . Here's the test cases for GET /organizations/{organization_id}/programs/{program_id}. Please review when you have time. 😉

File: tests/organizations/test_get_program.py [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+    @patch("requests.get")
+    def test_api_dao_get_program_successfully(self, mock_get_representative):
+        # prepare existing program with 
+        # start date 2020-10-15 12:00 GMT+0
+        # and end date 2020-11-25 12:00 GMT+0
+        program = ProgramModel(
+            program_name="Program A",
+            start_date=1602763200,
+            end_date=1606305600,
+            organization_id=self.organization1_data,
```
Comment (devkapilbansal) [inline tests/organizations/test_get_program.py]: Just want to ask why organization1_data is used instead of organization1_data.id

Comment (mtreacy002) [reply tests/organizations/test_get_program.py]: @KapilBansal320, this is because Organization to Program has one-to-many relationship where as other tables relation (user to organization, user to user_extension, user to personal_background) has one-to-one relationship. If you try put `id` behind `self.organization1_data`, you will get this error
Screen Shot 2020-09-16 at 11 12 10 am

and when you look up online, the cause was mentioned in the [stackoverflow discussion](https://stackoverflow.com/questions/16151729/attributeerror-int-object-has-no-attribute-sa-instance-state) that in one-to-many relation we want to have it as ORM object instead of a reference to PK.

**Comment by mtreacy002** [comment]
@mayburgos , @annabauza or @anitab-org/bridgeintech-maintainers , can someone please review this PR? Thanks

**Review by rpattath — DISMISSED** [review]
I ran the test locally and passed

File: tests/organizations/test_get_program.py [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+            "target_candidate_mental_ability": "Not Applicable",
+            "target_candidate_socio_economic": "Not Applicable",
+            "target_candidate_highest_education": "Not Applicable",
+            "target_candidate_years_of_experience": "Not Applicable",
+            "target_candidate_other": "Non-Binary",
+        }
+        program.contact_type = "REMOTE"
+        program.zone = "GLOBAL"
+        program.status = "DRAFT"
+        program.creation_date = 1601424000
```
Comment (isabelcosta) [inline tests/organizations/test_get_program.py]: can you use a code generated timestamp here please

File: tests/organizations/test_get_program.py [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+        organization2.save_to_db()
+        organization2_data = OrganizationModel.find_by_representative(test_user2_data.id)
+        
+        # prepare existing program with 
+        # start date 2020-10-15 12:00 GMT+0
+        # and end date 2020-11-25 12:00 GMT+0
+        program = ProgramModel(
+            program_name="Program A",
+            start_date=1602763200,
+            end_date=1606305600,
```
Comment (isabelcosta) [inline tests/organizations/test_get_program.py]: same comment as above :)

File: tests/organizations/test_get_program.py (hunk 2 above)
Comment (mtreacy002) [reply tests/organizations/test_get_program.py]: Done 😉

File: tests/organizations/test_get_program.py (hunk 3 above)
Comment (mtreacy002) [reply tests/organizations/test_get_program.py]: Done 😉

**Comment by mtreacy002** [comment]
@isabelcosta and @rpattath , I've made the requested changes. But this PR is dependent on PR #167 for the Travis to pass the other tests.

**Review by devkapilbansal — CHANGES_REQUESTED** [review]
All test cases passed locally but I found that there are some packages that are imported but never used. Also, there are some PEP8 guidelines formatting here.

TLDR
```
test_get_program.py:2:1: F401 'unittest' imported but unused
test_get_program.py:3:1: F401 'http.cookies' imported but unused
test_get_program.py:5:1: F401 'requests' imported but unused
test_get_program.py:6:1: F401 'requests.exceptions.HTTPError' imported but unused
test_get_program.py:10:1: F401 'app.database.sqlalchemy_extension.db' imported but unused
test_get_program.py:12:1: F401 'app.api.request_api_utils.post_request' imported but unused
test_get_program.py:12:1: F401 'app.api.request_api_utils.get_request' imported but unused
test_get_program.py:12:1: F401 'app.api.request_api_utils.BASE_MS_API_URL' imported but unused
test_get_program.py:12:80: E501 line too long (93 > 79 characters)
test_get_program.py:13:80: E501 line too long (96 > 79 characters)
test_get_program.py:14:1: F401 'app.api.models.organization.get_organization_response_model' imported but unused
test_get_program.py:14:1: F401 'app.api.models.organization.update_organization_request_model' imported but unused
test_get_program.py:14:80: E501 line too long (106 > 79 characters)
test_get_program.py:28:1: W293 blank line contains whitespace
test_get_program.py:31:80: E501 line too long (96 > 79 characters)
test_get_program.py:41:1: W293 blank line contains whitespace
test_get_program.py:48:1: W293 blank line contains whitespace
test_get_program.py:53:1: W293 blank line contains whitespace
test_get_program.py:55:13: F841 local variable 'login_response' is assigned to but never used
test_get_program.py:65:40: W291 trailing whitespace
test_get_program.py:66:34: W291 trailing whitespace
test_get_program.py:74:80: E501 line too long (80 > 79 characters)
test_get_program.py:85:44: W291 trailing whitespace
test_get_program.py:100:80: E501 line too long (99 > 79 characters)
test_get_program.py:101:1: W293 blank line contains whitespace
test_get_program.py:103:80: E501 line too long (113 > 79 characters)
test_get_program.py:105:1: W293 blank line contains whitespace
test_get_program.py:106:5: E303 too many blank lines (2)
test_get_program.py:108:40: W291 trailing whitespace
test_get_program.py:141:1: W293 blank line contains whitespace
test_get_program.py:143:1: W293 blank line contains whitespace
test_get_program.py:151:80: E501 line too long (93 > 79 characters)
test_get_program.py:152:80: E501 line too long (89 > 79 characters)
test_get_program.py:190:80: E501 line too long (99 > 79 characters)
test_get_program.py:193:1: W293 blank line contains whitespace
test_get_program.py:202:80: E501 line too long (87 > 79 characters)
test_get_program.py:211:1: W293 blank line contains whitespace
test_get_program.py:213:5: E303 too many blank lines (2)
test_get_program.py:231:1: W293 blank line contains whitespace
test_get_program.py:232:5: E303 too many blank lines (2)
test_get_program.py:233:80: E501 line too long (102 > 79 characters)
test_get_program.py:237:40: W291 trailing whitespace
test_get_program.py:238:34: W291 trailing whitespace
test_get_program.py:246:1: W293 blank line contains whitespace
test_get_program.py:249:39: W291 trailing whitespace
test_get_program.py:263:80: E501 line too long (89 > 79 characters)
test_get_program.py:264:1: W293 blank line contains whitespace
test_get_program.py:265:40: W291 trailing whitespace
test_get_program.py:270:1: W293 blank line contains whitespace
test_get_program.py:271:9: E303 too many blank lines (2)
test_get_program.py:297:1: W293 blank line contains whitespace
test_get_program.py:299:1: W293 blank line contains whitespace
test_get_program.py:306:1: W293 blank line contains whitespace
test_get_program.py:308:80: E501 line too long (82 > 79 characters)
```
Although the E501 line too long error can be ignored but I think it would be better if other changes are made.

**Comment by mtreacy002** [comment]
@devkapilbansal , can you please show me how you get the pylint report? Bcoz I tried to run pylint on the same file using the commit id `a162908` that I have on this PR it didn't give the same result as yours. Below are the screenshots:
* commit ID where the pylint was run:
Screen Shot 2020-10-22 at 2 41 26 pm

* running pylint under virtual env (the correct way):
Screen Shot 2020-10-22 at 2 41 48 pm

* even when I run pylint outside virtualenv on the same commit ID (not the correct way) I can't get the same result as yours:
Screen Shot 2020-10-22 at 2 43 15 pm

**Comment by mtreacy002** [comment]
PS: do you have the following .pylintrc on your code as what I have on this PR?
https://github.com/mtreacy002/bridge-in-tech-backend/blob/Test--GET-and-PUT-program-tests/.pylintrc

**Comment by mtreacy002** [comment]
Also, I'm curious to how all tests passed locally for you since for me majority of the tests resulted in error (since except for this PR file `test_get_program.py`, I didn't refactored the access_expiry for the token on this PR since they were done in PR #167 which this PR dependent upon and should be approved and merged beforehand).

**Comment by devkapilbansal** [comment]
> PS: do you have the following .pylintrc on your code as what I have on this PR?
> https://github.com/mtreacy002/bridge-in-tech-backend/blob/Test--GET-and-PUT-program-tests/.pylintrc

I use flask for this

**Comment by devkapilbansal** [comment]
> Also, I'm curious to how all tests passed locally for you since for me majority of the tests resulted in error (since except for this PR file `test_get_program.py`, I didn't refactored the access_expiry for the token on this PR since they were done in PR #167 which this PR dependent upon and should be approved and merged beforehand).

Tested it by changing time of system manually

**Review by isabelcosta — APPROVED** [review]
💯 @mtreacy002

**Comment by mtreacy002** [comment]
@rpattath , is it ok for me to merge this PR? @isabelcosta has approved this. It has been over 2 months since this PR was opened.

**Comment by isabelcosta** [comment]
@mtreacy002 I will merge this, from what I see in the comments the tests ran correctly.
Regarding the PEP8, I think this conversation can move to a separate issue, so these issues are fixed :)
